Use node lease

mshitrit opened the pull request:
As Part  of [ECOPROJECT-1310](https://issues.redhat.com//browse/ECOPROJECT-1310) implementing aquiring/releasing of node lease related to the creation/deletion of the remediations.
Design [doc](https://docs.google.com/document/d/1VAFDnXNd_yzAn2YX0I-thMe4tCtbXYIYU9QzirNJ8cY/edit#heading=h.s86s51b1rmwu)

########################################################
In order to make review easier I've tried to add relevant data on each commit.
########################################################

mshitrit commented:
/test ci/prow/4.13-openshift-e2e

mshitrit commented:
/ test 4.13-openshift-e2e

mshitrit commented:
/test 4.13-openshift-e2e

mshitrit commented:
/test 4.13-openshift-e2e

mshitrit commented:
/test 4.12-openshift-e2e

razo7 commented on go.mod:
```diff
@@ -11,14 +11,16 @@ require (
+	sigs.k8s.io/controller-runtime v0.14.5
 )
 
+require github.com/medik8s/common v1.0.0
```
NIT: Why adding this require here and not grouped it with the other packages from above (lines 6-19)?

razo7 replied:
I see it was modified in the [last commit](https://github.com/medik8s/node-healthcheck-operator/pull/212/commits/954af23f72c709a14111a250b69db0085400d0dc)....

mshitrit commented:
/test 4.13-openshift-e2e

clobrano commented on controllers/nodehealthcheck_controller_test.go:
```diff
@@ -428,6 +447,120 @@ var _ = Describe("Node Health Check CR", func() {
+								HaveField("Status", metav1.ConditionFalse),
+								HaveField("Reason", v1alpha1.ConditionReasonEnabled),
+							)))
+						//debugDelay()
```
Is this still necessary?

mshitrit commented on controllers/utils/utils.go:
```diff
@@ -51,12 +51,12 @@ func GetLogWithNHC(log logr.Logger, nhc *v1alpha1.NodeHealthCheck) logr.Logger {
+// MinRequeueDuration returns the minimal valid requeue duration
+func MinRequeueDuration(old, new *time.Duration) *time.Duration {
+	if new == nil || *new == 0 {
 		return old
 	}
-	if old == nil || *new < *old {
+	if old == nil || *old == 0 || *new < *old {
```
Yep, 0 isn't a valid value for requeueu (i.e 0 is the default value when there is no requeue delay)

clobrano replied:
what if also the other is `zero`? :sweat_smile:

mshitrit replied:
Than the remediation should not be delayed and the return value (0) would reflect that.

clobrano replied:
got it now, thanks

mshitrit commented on controllers/nodehealthcheck_controller_test.go:
```diff
@@ -428,6 +447,120 @@ var _ = Describe("Node Health Check CR", func() {
+						Expect(err).NotTo(HaveOccurred())
+					})
+
+					It("a remediation CR isn't created", func() {
```
Good point, although I'd phrase it a bit diffrently: `CR is created only when lease is obtained`.
(The lease can only be obtained by NHC only after the non-NHC node lease expires)

clobrano commented:
/lgtm
leaving time for more feedbacks
/hold

slintes commented on controllers/resources/lease.go:
```diff
@@ -0,0 +1,272 @@
+}
+
+// needs to release leases that needs to be released and extend those that needs extending return minimum timeout for next check
+func (m *nhcLeaseManager) manageLeases(ctx context.Context, nhc *remediationv1alpha1.NodeHealthCheck) (time.Duration, error) {
```
I didn't understand yet what this thing is doing, but iterating all nodes should be avoided IMHO. Will try give better feedback with idea what to change in the next round of review...

mshitrit replied:
In a nutshell, the nodes are used to find all potential leases that may exist, than each found lease is managed separately (i.e extended, deleted and getting the closest requeue)

slintes reviewed (changes requested):
I see you had dependency issues between the old resource manager and you new lease manager, which you resolved by making methods to functions. Not a very nice workaround IMHO. What about just integrating the lease methods into the existing resource manager, that would save some trouble IMHO?

mshitrit commented on controllers/resources/lease.go:
```diff
@@ -0,0 +1,272 @@
+	}, nil
+}
+
+func (m *nhcLeaseManager) ObtainNodeLease(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) (bool, *time.Duration, error) {
```
In a nutshell it makes it a bit harder to break the interface and creates a layer which logically separates interface form implementation (for example the private method can have a different signature without changing the interface).

I don't see it as a big deal and I'm happy to respect NHC house rules, so I'll gladly make the changes :+1:

mshitrit commented on controllers/nodehealthcheck_controller.go:
```diff
@@ -216,6 +223,12 @@ func (r *NodeHealthCheckReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 	}()
 
+	defer func() {
+		var leaseRequeue time.Duration
```
Sure, I was toying was idea myself but wasn't sure whether to do it or not

mshitrit commented on vendor/github.com/medik8s/common/pkg/lease/manager.go:
```diff
@@ -0,0 +1,308 @@
+	log.Info("request lease")
+	lease, err := l.getLease(ctx, obj)
+	//couldn't get the lease try to create one
+	if err != nil {
```
Yep, however this is part of common so I'll handle this in a different PR

mshitrit commented on vendor/github.com/medik8s/common/pkg/lease/manager.go:
```diff
@@ -0,0 +1,308 @@
+	} else {
+		// can't take over the lease if it is currently valid.
+		if isValidLease(lease, currentTime.Time) {
+			return fmt.Errorf("can't update valid lease held by different owner")
+		}
```
PR: https://github.com/medik8s/common/pull/3

mshitrit commented on controllers/resources/lease.go:
```diff
@@ -0,0 +1,272 @@
+	leaseDurationWithBuffer := leaseDuration + LeaseBuffer
+
+	node := &v1.Node{}
+	if err := m.client.Get(context.Background(), types.NamespacedName{Name: nodeName}, node); err != nil {
```
I don't quite follow, maybe you can show a code snippet or share a link of what you mean ?

slintes replied:
nevermind, my idea won't work, we need the UUID of the node for the lease

mshitrit replied:
On the one hand I do agree it would reserve unnecessary Get call, but on the other hand I don't think it's major here since it happens when a remediation is created (which does not happen very often) and we'll need to add the node to `CreateRemediationCR` which isn't very nice IMO.
I'm slightly inclined to leave it as is but Not feeling strongly either way, so let me know what you prefer.

slintes replied:
wfm

mshitrit commented on controllers/resources/lease.go:
```diff
@@ -0,0 +1,272 @@
+	}
+
+	if err := m.commonLeaseManager.RequestLease(context.Background(), node, leaseDurationWithBuffer); err != nil {
+		l, _ := m.commonLeaseManager.GetLease(context.Background(), node)
```
Yep, I'll update after https://github.com/medik8s/common/pull/3 is merged

mshitrit commented on controllers/nodehealthcheck_controller.go:
```diff
@@ -459,13 +472,22 @@ func (r *NodeHealthCheckReconciler) remediate(node *v1.Node, nhc *remediationv1a
 			return nil, nil
 		}
 		return nil, errors.Wrapf(err, "failed to create remediation CR")
+	} else {
```
It is true that the else block can be logically  be dropped, but IMO this sort of issue falls into code style category.
Arguably having  `if` and `else` blocks  in this example clearly defines the two separate use cases intended thus making the code more readable. 

In any case, this is very minor IMO so I don't mind making the change.

slintes replied:
I saw code linters (Codacity?) complaining about this 🤷🏼‍♂️

slintes replied:
see e.g.
https://go.dev/doc/effective_go#if
https://google.github.io/styleguide/go/decisions#indent-error-flow

mshitrit commented on controllers/resources/manager.go:
```diff
@@ -136,31 +141,44 @@ func (m *manager) GenerateRemediationCRBase(gvk schema.GroupVersionKind) *unstru
-	if err := m.Create(m.ctx, remediationCR); err != nil {
-		m.log.Error(err, "failed to create an external remediation object")
-		return false, err
+	isLeaseObtained, requeue, err := m.leaseManager.ObtainNodeLease(remediationCR, nhc)
```
Hmm I'm not sure.
On the one hand I do think you have a valid point that usually the requeue time is handled in `Reconcile` , however on the other hand the logic deciding the requeue time here is not trivial and has to do with leases so it does make sense that it'll be in the lease manager component.

If it's not something you feel strongly about I think I prefer to keep it as is.

mshitrit commented on controllers/nodehealthcheck_controller.go:
```diff
@@ -459,13 +472,22 @@ func (r *NodeHealthCheckReconciler) remediate(node *v1.Node, nhc *remediationv1a
 
 	// create remediation CR
-	created, err := rm.CreateRemediationCR(remediationCR, nhc)
+	created, leaseRequeueTimeout, err := rm.CreateRemediationCR(remediationCR, nhc)
```
>I think here it still is the leaseDuration

I guess it's kind of an egg an chicken dilemma, ATM I'm taking the approach that requeue time is managed in lease manager (vs in `Reconcile` method),  so this code reflects the idea that when a CR is created a requeue time is essential in order to manage that CR's lease (i.e extend it or remove it).

>requeueIn := leaseDuration (-someBuffer!?) )

Basically here it's the same concept as in the above comment, In the approach that I'm currently taking this part (i.e applying the buffer) is already managed in the lease manager so it would be transparent to the reconciler.

>(timeout?)

I've removed `timeout` suffix - I agree it's redundant here.

mshitrit commented on controllers/nodehealthcheck_controller.go:
```diff
@@ -459,13 +472,22 @@ func (r *NodeHealthCheckReconciler) remediate(node *v1.Node, nhc *remediationv1a
+		isLeaseObtained := leaseRequeueTimeout != nil && created
+		isCrAlreadyExist := !created && leaseRequeueTimeout == nil
+		isUnhealthyWithoutRemediation := !isLeaseObtained && !isCrAlreadyExist
+		//An unhealthy node exist but remediation couldn't be created because lease wasn't obtained - update unhealthy nodes and requeue
```
I'll take a crack at some refactoring after https://github.com/medik8s/common/pull/3 is merged

mshitrit commented:
> I see you had dependency issues between the old resource manager and you new lease manager, which you resolved by making methods to functions. 

That is true, basically the only method I needed was `ListRemediationCRs` and I didn't want to duplicate it's code.

>What about just integrating the lease methods into the existing resource manager

I don't think it's a good idea, I think it might have made sense for a slimer functionality - but in this use-case both are pretty big, so IMO this module separation would make the code more modular and easier to maintain.

>Not a very nice workaround IMHO.

Any particular reason that bothers you ? I might be missing something but I don't really see a problem with `ListRemediationCRs` being a function over a method.

In any case I can suggest two simple alternatives, let me know if any of them sounds better:

- restore manager `ListRemediationCRs`  method  and have it use listRemediationCRs function (basically keeping the shared code in a function without removing manager's method)
-  revert the change in manager (i.e make `ListRemediationCRs` a method again instead of `listRemediationCRs` function) and have  a method param in lease manager which will be set with manager's  method.
something like:
```
leaseManager.ListRemediationCRs = resourceManager.ListRemediationCRs
```

mshitrit commented on controllers/nodehealthcheck_controller.go:
```diff
@@ -194,12 +196,17 @@ func (r *NodeHealthCheckReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 	nhcOrig := nhc.DeepCopy()
 	var finalRequeueAfter *time.Duration
 	defer func() {
+		finalRequeueAfter = utils.MinRequeueDuration(&result.RequeueAfter, finalRequeueAfter)
```
I've also [mentioned](https://github.com/medik8s/node-healthcheck-operator/pull/215#discussion_r1230714051) this on your original PR, you've probably missed it due to my poor phrasing :see_no_evil:

slintes commented:
> > I see you had dependency issues between the old resource manager and you new lease manager, which you resolved by making methods to functions.
> 
> That is true, basically the only method I needed was `ListRemediationCRs` and I didn't want to duplicate it's code.
> 

duplicating code is as bad of course

> > What about just integrating the lease methods into the existing resource manager
> 
> I don't think it's a good idea, I think it might have made sense for a slimer functionality - but in this use-case both are pretty big, so IMO this module separation would make the code more modular and easier to maintain.
> 
> > Not a very nice workaround IMHO.
> 
> Any particular reason that bothers you ? I might be missing something but I don't really see a problem with `ListRemediationCRs` being a function over a method.

I  just think that breaking the concept of having managers with their own methods, by moving the implementatiom of a method into a package public function, is bad design. 

> 
> In any case I can suggest two simple alternatives, let me know if any of them sounds better:
> 
> * restore manager `ListRemediationCRs`  method  and have it use listRemediationCRs function (basically keeping the shared code in a function without removing manager's method)

Doesn't resolve my issue with the code

> * revert the change in manager (i.e make `ListRemediationCRs` a method again instead of `listRemediationCRs` function) and have  a method param in lease manager which will be set with manager's  method.
>   something like:
> 
> ```
> leaseManager.ListRemediationCRs = resourceManager.ListRemediationCRs
> ```

Hard to read and maintain IMHO, and it breaks the modularity you want to have.

What about making the CR list a parameter of `calculateLeaseDurationByAllRemediations()`, and let the reconciler call the resource managers method to fill it?

slintes commented on controllers/resources/lease.go:
```diff
@@ -0,0 +1,263 @@
+
+}
+
+func (m *nhcLeaseManager) UpdateReconcileResults(ctx context.Context, nhc *remediationv1alpha1.NodeHealthCheck, initialRequeue time.Duration, initialErr error) (time.Duration, error) {
```
This method is giving me headache. 
- we get a list of all nodes, which can be a potentially huge list
- then for each node, we try to get the lease
- all this while we most probably only need handle very few unhealthy node

Do we really need this?
Can't we do everything we need when iterating the (un)healhty nodes in the reconciler instead of the deferred update? E.g. `healthy node -> delete CR -> delete lease`. Or `unhealthy node -> CR already exists? -> check/extend lease duration`... :thinking:

mshitrit replied:
The rational was 
 - focus the lease logic in one place instead of spread it over the already crouded Reconcile code
 - "Better safe than sorry" approach which covers all possible leases.
However I do think you have a valid point regarding resource consumption - so no objection trying the alternative as well

mshitrit commented:
>Hard to read and maintain IMHO, and it breaks the modularity you want to have.

I agree it's a bit harder to read, not so much that it's an issue IMO, why do you think it breaks modularity  ?

Another option I had in mind is, instead of sharing the list code, set up only needed functionality (i.e ObtainNodeLease) in resource manager (instead of resource manager encapsulate lease manager).
So instead of:
```
	leaseManager, err := resources.NewLeaseManager(r.Client, log)
	if err != nil {
		return ctrl.Result{}, err
	}
	resourceManager := resources.NewManager(r.Client, ctx, r.Log, r.OnOpenShift, leaseManager)
```

Something like:
```
       resourceManager := resources.NewManager(r.Client, ctx, r.Log, r.OnOpenShift)
       leaseManager, err := resources.NewLeaseManager(r.Client, log, resourceManager)
       resourceManager.ObtainNodeLease = leaseManager.ObtainNodeLease
	
```
I realize it's share similarity with my previous suggestion of (`leaseManager.ListRemediationCRs = resourceManager.ListRemediationCRs`) but I thought there is no harm in suggesting.

>I just think that breaking the concept of having managers with their own methods, by moving the implementatiom of a method into a package public function, is bad design.

I'm not clear on why two managers sharing a common code is breaking a concept, what potential issues do you foresee as a result ? 


> What about making the CR list a parameter of `calculateLeaseDurationByAllRemediations()`, and let the reconciler call the resource managers method to fill it?

I'm not completely clear on what you mean, maybe share a short pseudo-code snippet ?

mshitrit commented on controllers/nodehealthcheck_controller.go:
```diff
@@ -459,7 +472,15 @@ func (r *NodeHealthCheckReconciler) remediate(node *v1.Node, nhc *remediationv1a
+
+	_, isLeaseAlreadyTaken := err.(*lease.AlreadyHeldError)
+	//An unhealthy node exist but remediation couldn't be created because lease wasn't obtained - update unhealthy nodes and requeue
+	if !!isLeaseAlreadyTaken {
```
Lol, no idea how I managed to get  that one in :smile:

mshitrit commented on main.go:
```diff
@@ -126,8 +126,9 @@ func main() {
+	client := mgr.GetClient()
 	if err := (&controllers.NodeHealthCheckReconciler{
-		Client:                      mgr.GetClient(),
+		Client:                      client,
```
good question, I assume in some point I've used `client` for controllers.MachineHealthCheckReconciler as well, but obviously it doesn't make sense as is - I'll revert it.

mshitrit commented on controllers/resources/lease.go:
```diff
@@ -0,0 +1,263 @@
+	return updatedRequeue, updatedErr
+}
+
+func (m *nhcLeaseManager) calculateLeaseDurationBySingleRemediation(remediationCR *unstructured.Unstructured, nhc *remediationv1alpha1.NodeHealthCheck) time.Duration {
```
>would getTimeoutForRemediation() be a more descriptive name?

Sure, that's fine by me

>And use the default at caller side when this returns 0?

I actually rather keep this in the current method, I think extracting it will cause less clean code. 
(I'll even remove the redundant else in `calculateLeaseDurationBySingleRemediation` :smile: )

slintes replied:
It's about separation of concerns, or modularity, or whatever it's called. And about meaningful names.
The thing the method is doing is: find the timeout. That's why I proposed the new name. But then the default lease duration has nothing to do with the timeout.
We could add an intermediate method:
```
getLeaseDurationForRemediation() {
  if  timeout := getTimeoutForRemediation(); timeout != 0 {
     return timeout
  }
  return defaultLeaseDuration
}
```
But IMHO this isn't worth it, it can be handled by the caller of getTimeoutForRemediation() directly.
Does that make sense? 🤔

slintes commented:
> > Hard to read and maintain IMHO, and it breaks the modularity you want to have.
> 
> I agree it's a bit harder to read, not so much that it's an issue IMO, why do you think it breaks modularity ?

with modularity I mean the concept of having managers with their own methods

> 
> Another option I had in mind is, instead of sharing the list code, set up only needed functionality (i.e ObtainNodeLease) in resource manager (instead of resource manager encapsulate lease manager). So instead of:
> 
> ```
> 	leaseManager, err := resources.NewLeaseManager(r.Client, log)
> 	if err != nil {
> 		return ctrl.Result{}, err
> 	}
> 	resourceManager := resources.NewManager(r.Client, ctx, r.Log, r.OnOpenShift, leaseManager)
> ```
> 
> Something like:
> 
> ```
>        resourceManager := resources.NewManager(r.Client, ctx, r.Log, r.OnOpenShift)
>        leaseManager, err := resources.NewLeaseManager(r.Client, log, resourceManager)
>        resourceManager.ObtainNodeLease = leaseManager.ObtainNodeLease
> 	
> ```

IMHO this `resourceManager.ObtainNodeLease = leaseManager.ObtainNodeLease` is very confusing

> 
> I realize it's share similarity with my previous suggestion of (`leaseManager.ListRemediationCRs = resourceManager.ListRemediationCRs`) but I thought there is no harm in suggesting.
> 
> > I just think that breaking the concept of having managers with their own methods, by moving the implementatiom of a method into a package public function, is bad design.
> 
> I'm not clear on why two managers sharing a common code is breaking a concept, what potential issues do you foresee as a result ?

Separation of concerns. The resource manager knows how to list CRs. When you move that into a package public function, that concept is broken.
One could even argue: why is the lease manager in the same package as the resource manager? When you would move it into its own package, the issue is clearer (it would need a global public function), is it?

> 
> > What about making the CR list a parameter of `calculateLeaseDurationByAllRemediations()`, and let the reconciler call the resource managers method to fill it?
> 
> I'm not completely clear on what you mean, maybe share a short pseudo-code snippet ?

pseudo code, in the reconciler:
crList := resourceManager.ListRemediationCRs()
leaseManager.UpdateReconcileResults(crList...)

Btw, that would also prevent to call listRemediationCRs potentially 3 times for every lease in manageLease... (in isOverdue, isRemediationExists and calculateLeaseDurationByAllRemediations)

mshitrit commented:
>pseudo code, in the reconciler:
crList := resourceManager.ListRemediationCRs()
leaseManager.UpdateReconcileResults(crList...)

>Btw, that would also prevent to call listRemediationCRs potentially 3 times for every lease in manageLease... (in isOverdue, isRemediationExists and calculateLeaseDurationByAllRemediations)

Sounds good, WFM :+1:

mshitrit commented on controllers/nodehealthcheck_controller.go:
```diff
@@ -194,12 +198,21 @@ func (r *NodeHealthCheckReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 	var finalRequeueAfter *time.Duration
 	defer func() {
+		var leaseRequeue time.Duration
+		leaseRequeue, returnErr = leaseManager.UpdateReconcileResults(ctx, nhc, result.RequeueAfter, returnErr)
```
This probably not relevant anymore after refactoring

slintes commented on controllers/resources/manager.go:
```diff
@@ -136,55 +142,87 @@ func (m *manager) GenerateRemediationCRBase(gvk schema.GroupVersionKind) *unstru
+	defer func() {
+		if isDeleted {
+			//make sure lease will be deleted
+			remediationCrs = nil
```
this is hard to understand... why do we set remediationCrs to nil here, and wy do we need them at all? :thinking:

mshitrit replied:
We fetch the remediation before the remediation is deleted, so `remediationCrs` shows a state which is no longer correct and will cause ManageLease to produce wrong output.

slintes reviewed (changes requested):
This looks better, but I still don't understand everything. Especially the ManagerLease method is a mystery. It's doing things it doesn't need to do IMHO, just because it doesn't know why it's called, it is too generic.

- Why is it called when a remediation CR for a unheathy node already exists? This usecase (potentially extending a lease) can be covered by ObtainLease (which calls `commonLeaseManager.RequestLease()`), not?

- When it's called for a healthy node, it just needs to invalidate the lease, not?

mshitrit commented on controllers/resources/manager.go:
```diff
@@ -136,55 +142,87 @@ func (m *manager) GenerateRemediationCRBase(gvk schema.GroupVersionKind) *unstru
 		}
 		m.log.Info("external remediation CR already exists", "CR name", remediationCR.GetName(), "kind", remediationCR.GetKind(), "namespace", remediationCR.GetNamespace())
-		return false, nil
+		remediationCrs, err := m.ListRemediationCRs(nhc, func(cr unstructured.Unstructured) bool {
```
In this use case the CR already exist therefore a lease most likely exist as well, this lease may need to be extended or removed

mshitrit commented on controllers/nodehealthcheck_controller.go:
```diff
@@ -312,7 +321,9 @@ func (r *NodeHealthCheckReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 		}
 		for _, remediationCR := range remediationCRs {
-			if deleted, err := resourceManager.DeleteRemediationCR(&remediationCR, nhc); err != nil {
+			deleted, leaseRequeueIn, err := resourceManager.DeleteRemediationCR(&remediationCR, nhc, remediationCRs)
```
>my initial thought when seeing this is: when I delete a CR, how can this result in a requeue?
Delete is not necessarily successful, IIUC in case it's not, requeue makes sense

mshitrit replied:
>also, why does it need the complete CR list? Maybe I miss something, but this looks very unintuitive to me thinking

Initially it was done so  the shortest requeue will be selected among remediations, but I agree you are correct - I'll update the code

slintes replied:
> Delete is not necessarily successful, IIUC in case it's not, requeue makes sense

in this case an error should be returned

mshitrit commented:
> * Why is it called when a remediation CR for a unheathy node already exists? This usecase (potentially extending a lease) can be covered by ObtainLease (which calls `commonLeaseManager.RequestLease()`), not?

Not necessarily,  the lease may also be needed to be deleted `ManageLease` contains and applies the logic of whether the lease would be extended or deleted.

> * When it's called for a healthy node, it just needs to invalidate the lease, not?

ATM we've modified the logic so it'll be called upon Creation/Deletion of CR so I'm not sure if `called for a healthy node` is the correct term here.
It is true that for most cases for a deleted CR the lease would be invalidated (i.e when the CR is successfully deleted)

>It's doing things it doesn't need to do IMHO, just because it doesn't know why it's called, it is too generic.

I agree that in theory we can export some of the logic outside of ManageLease and simplify it, but that would be at the expense of moving lease logic into "non lease" files, I advice against that:
- I think that we should strive to have  lease code in lease manager
- IMO ManageLease is pretty clear in what it does - delete or extend leases according to the need (i.e manages the lease)

slintes commented:
> After a lease for an unhealthy node is extended it could also be deleted

ah ok, the overdue part

> I don't see a reason for adding another method.

it's not about another method, but putting things where they better belong to IMHO

quoting myself:

> everything for unhealhy nodes can be in ObtainNodeLease, which uses commonLeaseManager.RequestLease (maybe some parts of the current ManageLease method need to go here, or maybe even in to the common manager?)

> everything for healthy nodes can be in a ReleaseLease method, which uses commonLeaseManager.InvalidateLease

maybe code can better explain what I'm aiming for, see: https://github.com/slintes/node-healthcheck-operator/commit/e9f88ef6f5d820a5003815fb90125368849787cd

- I renamed `ObtainNodeLease` to `RequestOrExtendNodeLease`, and that handles unhealthy nodes
- Relevant parts of `ManageLease` went into above method. I think they can even go into common lease manager, see suggestions in TODOs in the code
- Leftover in `ManageLease` is just lease invalidation, which is called for healthy nodes after CR was deleted, so I renamed it to just that, `InvalidateLease`

I noticed we miss something so far: what to do with the remediation CR when a lease is overdue and invalidated. Delete? Timeout?

mshitrit commented:
>I renamed ObtainNodeLease to RequestOrExtendNodeLease, and that handles unhealthy nodes

I don't think there is any value for renaming here, IMO names are equivalent - I do think that ObtainNodeLease is a better name (both names describe the action of getting a valid lease but IMO `ObtainNodeLease` is more concise, so a better name).
Not a major issue IMO I can do with  `RequestOrExtendNodeLease` if you feel strongly about it.

Having `RequestOrExtendNodeLease` deleting leases (for overdue lease) is just wrong IMO, it's not something I'd expect from a method with that name, a better name for a method that either extend / delete lease would be `ManageLease`.

>Leftover in ManageLease is just lease invalidation, which is called for healthy nodes after CR was deleted, so I renamed it to just that, InvalidateLease

I'm fine with using (and exposing) InvalidateLease after CR was deleted, I agree it might be a bit clearer - I think there is a bit of downside of  exposing another API, but I'm ok with that trade-off.

I think that having `remediate` method deal with `leaseOverdue` use case is wrong - it's better be handled withing lease module (i.e as now in ManageLease).

>I noticed we miss something so far: what to do with the remediation CR when a lease is overdue and invalidated. Delete? Timeout?


ATM CR is not deleted - deleting the CR when the node is still unhealthy seems like a bit of drastic measure for a first step. 
- in case we are at escalation use-case, since both use timeout the CR should be marked by NHC with the timeout annotation before the lease is deleted - of course we don't have 100% guarantee that remediators will be done with the CR, but I think it's an improvement over current situation.
- for non escalation,  we have a bit of buffer (~ 30 min) which which we can hope that in case the remediator is not done by than it's done working - of course this is not guaranteed, but IMO also improvement over current situation.

slintes commented:
> > I renamed ObtainNodeLease to RequestOrExtendNodeLease, and that handles unhealthy nodes
> 
> I don't think there is any value for renaming here, IMO names are equivalent - I do think that ObtainNodeLease is a better name (both names describe the action of getting a valid lease but IMO `ObtainNodeLease` is more concise, so a better name). Not a major issue IMO I can do with `RequestOrExtendNodeLease` if you feel strongly about it.
> 

I was just aligning to the common lease managers method name, which is request and not obtain 🤷🏼‍♂️ I think it would be more consistent. 

> Having `RequestOrExtendNodeLease` deleting leases (for overdue lease) is just wrong IMO, it's not something I'd expect from a method with that name, a better name for a method that either extend / delete lease would be `ManageLease`.
> 

fair point, maybe it would be best to have 3 methods:

- requestLease (or obtain)
- extendOrInvalidateLease
- invalidateLease

> > Leftover in ManageLease is just lease invalidation, which is called for healthy nodes after CR was deleted, so I renamed it to just that, InvalidateLease
> 
> I'm fine with using (and exposing) InvalidateLease after CR was deleted, I agree it might be a bit clearer - I think there is a bit of downside of exposing another API, but I'm ok with that trade-off.
> 

what do you mean with exposing another API?

> I think that having `remediate` method deal with `leaseOverdue` use case is wrong - it's better be handled withing lease module (i.e as now in ManageLease).

here I strongly disagree: the lease manager should just deal with leases, nothing more. When it starts dealing with remediations, we mix responsibilities, which makes the code harder to read, maintain and also test.
 
> 
> > I noticed we miss something so far: what to do with the remediation CR when a lease is overdue and invalidated. Delete? Timeout?
> 
> ATM CR is not deleted - deleting the CR when the node is still unhealthy seems like a bit of drastic measure for a first step.

agree

> 
> * in case we are at escalation use-case, since both use timeout the CR should be marked by NHC with the timeout annotation before the lease is deleted - of course we don't have 100% guarantee that remediators will be done with the CR, but I think it's an improvement over current situation.

I think order (lease invalidation / remediation timeout) matters only if we add some buffer here, by making the overdue process a 2 step process, e.g.:

- lease is overdue -> set timeout, requeue
- lease is overdue + some buffer -> invalidate lease 

> * for non escalation,  we have a bit of buffer (~ 30 min) which which we can hope that in case the remediator is not done by than it's done working - of course this is not guaranteed, but IMO also improvement over current situation.

where is that 30 minutes coming from? 🤔

mshitrit commented:
>fair point, maybe it would be best to have 3 methods:
requestLease (or obtain)
extendOrInvalidateLease
invalidateLease

I don't think it's fundamentally different from what we have now.
requestLease  is ObtainLease
extendOrInvalidateLease is ManageLease

>what do you mean with exposing another API?

Adding another method to LeaseManager `interface`

>where is that 30 minutes coming from? thinking

[Initial](https://github.com/medik8s/node-healthcheck-operator/pull/212/files#diff-b0df20bbf7681fe87317f506104c06107e4979f97a9e3698289a010890602672R29) 10 Minutes + [2 extensions](https://github.com/medik8s/node-healthcheck-operator/pull/212/files#diff-b0df20bbf7681fe87317f506104c06107e4979f97a9e3698289a010890602672R31) (each another 10 minutes)

>here I strongly disagree: the lease manager should just deal with leases, nothing more. When it starts dealing with remediations, we mix responsibilities, which makes the code harder to read, maintain and also test.

No objection to your statement but I don't think this is the case, it might have been the case if we needed a special treatment for a remediation for overdue leases (for example deleting them), but at the moment it just unnecessarily complicate  the code of  the `remediation` method by addressing this use case. 

>I think order (lease invalidation / remediation timeout) matters only if we add some buffer here, by making the overdue process a 2 step process, e.g.:

I might be misunderstanding you intention,  but  I agree a fix is needed here.
ATM the maximum duration before lease is considered overdue  would be timeout of current remediation X 3.
I think it should be (sum of previous timeouts until current one) + current one X 3

So for example:
for a lease that was created at 10:00, with the following remediations:

`Name`                 | `Order` | `timeout`
FastRemediation  |   0            | 2m
MedRemeiation    |   1            | 3m
SlowRemediation |   2            | 4m

- when processing `FastRemediation`  lease would expire at ~ 10:06 (same as now)
- when processing `MedRemeiation`    lease would expire at ~ 10:11 (2 + 3*3) , instead of 10:09 (currently)
- when processing `SlowRemediation` lease would expire at ~ 10:17 (2+3 + 3*4) , instead of 10:12 (currently)
Does that make sense ?

slintes commented:
> My idea is (for the single or last remediation): mark CR with timeout, wait a bit to let the remediator stop remediation and do potential cleanup, and only then invalidate lease. When there is another lease to try, just mark CR with timeout and immediately extend the lease.

but I'm fine with just invalidating immediately, return the Overdue error, and let the reconciler set the timeout annotation. Having an additional buffer probably has low value

mshitrit commented:
>no, the difference is that extendOrInvalidateLease is called for unhealthy nodes with the goal to extend the lease. Invalidation only happens when extending should not done anymore. ManageLease is also used for healthy nodes for just invalidating which makes it more complicated.

I don't think this statement is correct in the current code iteration (even though indeed it used to be correct).
ATM  `ManageLease` is only called [here](https://github.com/medik8s/node-healthcheck-operator/pull/212/files#diff-989521381369a7c73f553158810a1edd643d3b44d8cafdeb0da959dfbab4c7dfR154) for existing CR (i.e unhealthy node), and for healthy nodes (i.e when CR is deleted) `InvalidateLease` is called [here](https://github.com/medik8s/node-healthcheck-operator/pull/212/files#diff-989521381369a7c73f553158810a1edd643d3b44d8cafdeb0da959dfbab4c7dfR207)

>But that's what we need, handling the overdue lease case.

Maybe I'm missing something here.
The way I see it the reconciler wouldn't do anything different with the knowledge of a lease being overdue (so I don't necessarily consider it an error) and see no value of passing information I consider redundant to the reconciler.
I completely agree that in the future this information might be relevant so it would make sense to pass it along than.

Am I wrong in assuming this information redundant now ? , Is there a definite difference in the expected behavior ? 
IIUC In any case the lease would be deleted and reconcile will finish successfully without a requeue.

>My idea is (for the single or last remediation): mark CR with timeout

ATM the logic regarding placing the nhc-timeout annotation works separately form the lease logic (even though they do share a common thread of checking the CR's timeout).
in case we want to make changes on if/when to apply  nhc-timeout annotation for the last/single element I think it would be best to do it in a separate PR.

mshitrit commented:
/retest

slintes commented:
> > no, the difference is that extendOrInvalidateLease is called for unhealthy nodes with the goal to extend the lease. Invalidation only happens when extending should not done anymore. ManageLease is also used for healthy nodes for just invalidating which makes it more complicated.
> 
> I don't think this statement is correct in the current code iteration (even though indeed it used to be correct). ATM `ManageLease` is only called [here](https://github.com/medik8s/node-healthcheck-operator/pull/212/files#diff-989521381369a7c73f553158810a1edd643d3b44d8cafdeb0da959dfbab4c7dfR154) for existing CR (i.e unhealthy node), and for healthy nodes (i.e when CR is deleted) `InvalidateLease` is called [here](https://github.com/medik8s/node-healthcheck-operator/pull/212/files#diff-989521381369a7c73f553158810a1edd643d3b44d8cafdeb0da959dfbab4c7dfR207)
> 

oooh.... sorry, I missed your latest changes indeed, I didn't expect that you add the 3rd method while argueing against it 😉 

> > But that's what we need, handling the overdue lease case.
> 
> Maybe I'm missing something here. The way I see it the reconciler wouldn't do anything different with the knowledge of a lease being overdue (so I don't necessarily consider it an error) and see no value of passing information I consider redundant to the reconciler. I completely agree that in the future this information might be relevant so it would make sense to pass it along than.
> 
> Am I wrong in assuming this information redundant now ? , Is there a definite difference in the expected behavior ? IIUC In any case the lease would be deleted and reconcile will finish successfully without a requeue.
> 

I think we shouldn't ignore an overdue and invalidated lease. IMHO we need to stop remediation with the timeout annotation. And, while thinking more about it, it's not just the annotation for the remediator, we probably also need the timeout in our own status in order to avoid just creating a new lease on next reconcile... 🤔 

> > My idea is (for the single or last remediation): mark CR with timeout
> 
> ATM the logic regarding placing the nhc-timeout annotation works separately form the lease logic

that's why I suggested the LeaseOverDue error, that's IMHO the easiest way to connect the dots while keeping separation of concerns 

>  (even though they do share a common thread of checking the CR's timeout). in case we want to make changes on if/when to apply nhc-timeout annotation for the last/single element I think it would be best to do it in a separate PR.

I think we need it on this PR, see comment above

mshitrit commented:
>that's why I suggested the LeaseOverDue error, that's IMHO the easiest way to connect the dots while keeping separation of concerns

I still think it should belong in a different PR, but at least I agree that the required changes makes sense, so no strong objection to do it here.

>we probably also need the timeout in our own status in order to avoid just creating a new lease on next reconcile... thinking

I'm not completely clear on your intention here, maybe elaborate a bit more or add a snippet ?

>oooh.... sorry, I missed your latest changes indeed, I didn't expect that you add the 3rd method while argueing against it wink

Nah I was ok with the tradeoff (as I see it) :
>I'm fine with using (and exposing) InvalidateLease after CR was deleted, I agree it might be a bit clearer - I think there is a bit of downside of exposing another API, but I'm ok with that trade-off.

slintes commented:
> > we probably also need the timeout in our own status in order to avoid just creating a new lease on next reconcile... thinking
> 
> I'm not completely clear on your intention here, maybe elaborate a bit more or add a snippet ?
>

When we run into overdue, the lease is deleted. On the next reconcile we would just create a new lease, when we don't track the overdue/timeout in the NHC CR, not?
 
> > oooh.... sorry, I missed your latest changes indeed, I didn't expect that you add the 3rd method while argueing against it wink
> 
> Nah I was ok with what I consider a tradeoff in that case :
> 
> > I'm fine with using (and exposing) InvalidateLease after CR was deleted, I agree it might be a bit clearer - I think there is a bit of downside of exposing another API, but I'm ok with that trade-off.

fair enough, sorry :)

beekhof commented on controllers/resources/lease.go:
```diff
@@ -0,0 +1,249 @@
+		m.log.Info("managing lease - lease has no remediations so  about to be removed", "lease name", l.Name)
+		//release the lease - no remediations
+		return 0, m.commonLeaseManager.InvalidateLease(ctx, node)
+	} else if ok, err := m.isLeaseOverdue(l, nhc, remediationCR); err != nil {
```
My turn for a nit :-)

The `else`'s are redundant when every block returns

mshitrit replied:
I agree about the last `else`- it's not that small and we can cut off one nesting level by doing that

regarding the `else if's` I prefer to keep as is:
- I don't think they add little to no complexity in the code (over the alternative of just `if`)
- using that patter allow defining the variables withing the cond block and using it across all cases - IMO it is much nicer than having them predefined  (especially for such a use case with very small blocks)

beekhof commented on controllers/resources/lease.go:
```diff
@@ -0,0 +1,249 @@
+	return len(remediationCrs) > 0
+}
+
+func (m *nhcLeaseManager) logManageLeaseChanges(originalRequeue time.Duration, updatedRequeue time.Duration, originalErr error, updatedErr error) {
```
I couldn't find this being used. Did I miss something?

mshitrit replied:
No, good catch this is a leftover that was used in the old pattern where Reconcile iterate on nodes and Manage each  of them.

beekhof commented:
IMHO, what `ManageLease` does is fine, but it's being called from the wrong level.
The functions in `lease.go` should be being called directly from `Reconcile()` - being the place to co-ordinate the different modules.  I could also see `remediate()` being a viable alternative, eg.

```
created, err := rm.CreateRemediationCR(remediationCR, nhc)
if err == nil {
    duration, err := rm.leaseManager.ManageLease(rm.ctx, remediationCR, nhc)
}
...
```

This would also be consistent with exposing an `InvalidateLease` function for use when deleting CRs.

mshitrit commented:
>When we run into overdue, the lease is deleted. On the next reconcile we would just create a new lease, when we don't track the overdue/timeout in the NHC CR, not?

I don't think so.
- `ManageLease` would only extend existing lease (will not create a new one)
- `ObtainNodeLease` will not be called for an already existing CR, so it'll not create a lease for this use case as well.

slintes commented on controllers/resources/lease.go:
```diff
@@ -0,0 +1,233 @@
+	var sumDurations time.Duration
+	for _, esRemediation := range nhc.Spec.EscalatingRemediations {
+		if currentEscalatingRemediation.Order > esRemediation.Order {
+			sumDurations += esRemediation.Timeout.Duration
```
Just for the record, this calculation isn't accurate when a remediation failed early by setting the succeeded condition to false. But good enough, expiring later than intended isn't an issue.

slintes commented on controllers/resources/manager.go:
```diff
@@ -136,35 +137,53 @@ func (m *manager) GenerateRemediationCRBase(gvk schema.GroupVersionKind) *unstru
 		}
 		m.log.Info("external remediation CR already exists", "CR name", remediationCR.GetName(), "kind", remediationCR.GetKind(), "namespace", remediationCR.GetNamespace())
-		return false, nil
+		if err != nil {
```
redundant: err can't be != nil in this block, and that case is handled in the else block below

slintes commented on controllers/nodehealthcheck_controller.go:
```diff
@@ -459,7 +469,20 @@ func (r *NodeHealthCheckReconciler) remediate(node *v1.Node, nhc *remediationv1a
-	created, err := rm.CreateRemediationCR(remediationCR, nhc)
+	created, leaseRequeueIn, err := rm.CreateRemediationCR(remediationCR, nhc)
+
+	//An unhealthy node exist but remediation couldn't be created because lease wasn't obtained - update unhealthy nodes and requeue
```
this and below can go into the `if err != nil` block further down

slintes commented on controllers/nodehealthcheck_controller.go:
```diff
@@ -459,7 +469,20 @@ func (r *NodeHealthCheckReconciler) remediate(node *v1.Node, nhc *remediationv1a
+	//Lease is overdue
+	if _, isLeaseOverDue := err.(resources.LeaseOverDueError); isLeaseOverDue {
+		resources.UpdateStatusNodeUnhealthy(node, nhc)
+		return nil, r.addTimeOutAnnotation(rm, remediationCR, metav1.Time{Time: currentTime()})
```
I think we should update the NHC CR status here as well, similar to the `startedRemediation.TimedOut = &now` at the end of this method. Copying the `startedRemediation := ...` call to here should be fine, moving it above this block won't work.

slintes reviewed (commented):
I think we are getting to the finish line, some comments inline

mshitrit commented on controllers/resources/lease.go:
```diff
@@ -0,0 +1,233 @@
+	}
+	l, err := m.commonLeaseManager.GetLease(ctx, node)
+	if err != nil {
+		if errors.IsNotFound(err) {
```
>e.g. get a new one

why ? in case lease doesn't exist it would be because it's deleted because it's overdue

slintes replied:
anyone can delete the lease 🤷🏼‍♂️ 🙂

mshitrit replied:
IIUC we can abandon this thread, since we are already discussing this in the [thread below](https://github.com/medik8s/node-healthcheck-operator/pull/212#discussion_r1267886016.)

slintes commented on controllers/resources/lease.go:
```diff
@@ -0,0 +1,233 @@
+		return 0, err
+	}
+	exist := remediationCR.GetDeletionTimestamp() == nil
+	if !exist && m.isLeaseOwner(l) {
```
> The !exist doesn't check the lease, it checks whether the remediation is about to be deleted

Oh sorry, I misread the code 🙈 

> ManageLease is called only after a remediation CR is created, and the condition to creating a Remediation CR is owning the lease.

Anyone can change leases. You shouldn't update/delete not owned leases, but no one is preventing it. So IMHO better safe than sorry, let's not be that bad guy ourself, and always check ownership IMHO.

mshitrit replied:
Well consider the following scenario:
1. Unhealthy Node detected
2. Obtain lease creates a lease
3. Remediation is created
4. Node stays unhealthy for a long time (remediation is unsuccessful)
5. Lease is deleted because it's overdue.

Assuming we allow `ManageLease` to create leases: at that point (in case a remediation occur) we will create a new lease for that remediation, which would be wrong IMO.

It is true that technically active leases can be deleted by anyone, but that would "break the lease contract" so I think it's a reasonable assumption to assume this "contract" is upheld - I prefer this approach over the alternative I've described.

slintes replied:
>  at that point (in case a remediation occur) we will create a new lease for that remediation, which would be wrong IMO.

yes, that would be wrong, but I think that won't happen because we won't try to create a new remediation CR / lease for a timed out remediation

> Coming to think of it NMO original implementation of invalidating the lease didn't delete the lease, maybe it was so to address this type of use-case

might be... but leaving leases forever also is wrong, not? 🤔

mshitrit replied:
>yes, that would be wrong, but I think that won't happen because we won't try to create a new remediation CR / lease for a timed out remediation

We don't have to.
IIUC Any triggering to `Reconcile` (for example creating a remediation for another node) will cause that scenario since `ManageLease` is triggered for every unhealthy node.

>might be... but leaving leases forever also is wrong, not? thinking

That's a really good question.
On the hand it's removing the leases is definitely better way for cleanup & resource management, but on the other hand I don't think it's a huge issue leaving the lease since we still fulfill the "lease contract"  by "closing" it even though it's not deleted.

Thinking from the top of my mind: maybe a better solution will be to do the deletion of the lease only when the CR is removed :thinking:  (Not in this PR though :smile: )

slintes replied:
> IIUC Any triggering to Reconcile (for example creating a remediation for another node) will cause that scenario since ManageLease is triggered for every unhealthy node.

Ah yes, correct for the single remediation. Oh, and when setting timeout annotation / status update fails...

So the correct flow would be:

- in ManageLease:
  - for existing CR without timeout annotation:
    - when lease overdue: do not invalidate yet, just return OverDue error
    - when lease not found: create new
  - for existing CRs with timeout annotation: invalidate lease and return OverDue error (or a new error)
- in reconciler, when OverDue error occurs:
  - if timeout annotation does not exist: add it and requeue
  - if timeout annotation does exist:  done

Does that make sense? 🤔 

But we can implement this in a follow up, and skip creating new leases in ManageLease for now.

mshitrit replied:
>Ah yes, correct for the single remediation.

Why only single ? IIUC it also apply for escalating remediation - IIRC we do not delete the CR in escalating remediation so it will still be detected as unhealthy.

>But we can implement this in a follow up, and skip creating new leases in ManageLease for now.

IIUC this discussion is about better handling the use-case where someone has deleted a valid lease, I agree it's niche enough to discuss later on.

slintes replied:
hm, actually I think we need to apply parts of the flow, even if we don't want to create new leases:

when a lease is overdue and invalidated, but the remediation CR (timeout annotation) or NHC CR (status) update fails, these updates will not be tried again, because in the next reconcile ManageLease doesn't return the OverDue error again... 🤔

slintes replied:
> Why only single ?

because for escalating remediation it won't find an appropriate template which for which we doesn't have a timed out CR already, so it stops reconcile before trying to create a remediation CR

mshitrit replied:
> because in the next reconcile ManageLease doesn't return the OverDue error again... thinking

IIUC the TimeOutAnnotation will still be placed since it's checked independently of the overdue error [here](https://github.com/medik8s/node-healthcheck-operator/pull/212/files#diff-42b14dc2a70877933db4647434fcbe9b65a8859c7ec1911b0c9da7361c8e98b1R571).

>because for escalating remediation it won't find an appropriate template which for which we don't have a timed out CR already, so it stops reconcile before trying to create a remediation CR


Going back to this scenario: (let's say it happens for node A)
>1. Unhealthy Node detected
>2. Obtain lease creates a lease
>3. Remediation is created 
>4. Node stays unhealthy for a long time (remediation is unsuccessful)
>5. Lease is deleted because it's overdue.

- I think we can agree 1-5 can happen both in escalation / non escalation.
- Now let's say Node B becomes unhealthy and reconcile is triggered.
- we iterate over all existing remediation (in all nodes)
- Node A still has an existing remediation (might be even more than one)
- assuming manage lease also creates leases
- we will try to create a lease for node A, even though it already had a lease that was deleted 
- In this case a lease is created wrongfully because of a remediation that was created by escalation remediation 

Am I missing something here ?

slintes replied:
> IIUC the TimeOutAnnotation will still be placed since it's checked independently of the overdue error

hm, fair point, because overdue and timeout are directly related

> we will try to create a lease for node A, even though it already had a lease that was deleted

no, because `ManageLease()` won't be called for node A. When the lease was overdue, all possible remediations timed out, and `GetCurrentTemplateWithTimeout()` returns a `NoTemplateLeftError`, and stops remediation for that node.

mshitrit replied:
>all possible remediations timed out, and GetCurrentTemplateWithTimeout() returns a NoTemplateLeftError

Cool, thanks :+1:  - I was missing that part.

slintes commented on controllers/resources/manager.go:
```diff
@@ -161,7 +161,8 @@ func (m *manager) CreateRemediationCR(remediationCR *unstructured.Unstructured,
 
-	isLeaseObtained, requeue, err := m.leaseManager.ObtainNodeLease(remediationCR, nhc)
+	requeue, err := m.leaseManager.ObtainNodeLease(remediationCR, nhc)
+	isLeaseObtained := err == nil
```
I think we can just remove the `isLeaseObtained` var. After the error check below we know it was obtained, so the next `if` can be removed and the return value just be `true`

slintes commented on controllers/nodehealthcheck_controller.go:
```diff
@@ -481,7 +481,16 @@ func (r *NodeHealthCheckReconciler) remediate(node *v1.Node, nhc *remediationv1a
+			startedRemediation := resources.FindStatusRemediation(node, nhc, func(r *remediationv1alpha1.Remediation) bool {
+				return r.Resource.GroupVersionKind() == remediationCR.GroupVersionKind()
+			})
+			// update status (important to do this after CR update, else we won't retry that update in case of error)
```
There probably is a very low chance only that we don't find the right status, but just to be on the safe side I would also copy the nil check?
nit: also, it would be nice to use the same timestamp for the timeout annotation and the status

slintes commented on controllers/resources/lease.go:
```diff
@@ -108,8 +108,12 @@ func (m *nhcLeaseManager) ManageLease(ctx context.Context, remediationCR *unstru
 	}
+	//nothing to do with this lease
+	if !m.isLeaseOwner(nodeLease) {
+		return 0, nil
```
We are not owner, but we don't know if the lease is expired. So I think we can try to take the lease IMHO.
Same goes for the isNotFound case above, just try to create a new one, as mentioned in an earlier comment.

example: https://github.com/slintes/node-healthcheck-operator/commit/e9f88ef6f5d820a5003815fb90125368849787cd#diff-b0df20bbf7681fe87317f506104c06107e4979f97a9e3698289a010890602672R85

mshitrit replied:
`ManageLease` only extends or removes expired leases (i.e it manages existing leases).
In case we need to take a lease (whether we are the owner or not) it will happen in `ObtainLease`

slintes replied:
`ManageLease` is handling leases for existing remediation CRs. IMHO this doesn't exclude creating a new lease if needed.

mshitrit replied:
IIUC we can abandon this thread, since we are already discussing the term on which a lease will be created in the [thread above](https://github.com/medik8s/node-healthcheck-operator/pull/212#discussion_r1267886016).

slintes commented:
Looks good to me, I think we have pretty good version now 🙂

Thanks for you patience with my reviews! I know I can be picky, but a clear architecture and good code quality are important to me, for easier maintenance. Also, when new APIs enter the game, even when not public for users, it's important to start with a stable version IMHO, because updates to a new version are easy in code, but can potentially cause trouble when updating the running operator in a cluster 🙂

Adding the tests early was very valuable during refactorings, much appreciated!

I'm only aware of one potentially useful follow up for covering edge cases: in `ManageLease()` we can request a new lease when the old one was deleted, or we can take over a lease when it's not owned by us but expired.

/lgtm

slintes commented:
/hold cancel
